Fix s3 open()

rikigigi opened the pull request:
"file exist" operation now is performed with boto3's head_object
before it was using list_object, that by default returns 1000 objects at maximum. The requested (existing) key may not be in the returned list. In any case the maximum number of objects must be known in advance, and listing all objects is a waste of resources.

solves https://github.com/sphuber/aiida-s3/issues/11

Some points to discuss on my side:
 - call or not call super().open(key)? In principle I think it should be called, but the additional test on file existence is a waste of communication. If this line is removed this implementation works in the same way and is more efficient.
 - we can possibly find a way to implement list_objects to list all the objects in the bucket, but this may be useless and expensive for the server. Is this feature needed anywhere?

sphuber commented on src/aiida_s3/repository/aws_s3.py:
```diff
@@ -128,10 +135,14 @@ def open(self, key: str) -> t.Iterator[t.IO[bytes]]:  # type: ignore[override]
         :raise OSError: if the file could not be opened.
         """
         super().open(key)
```
```suggestion
```

sphuber reviewed (changes requested):
Thanks @rikigigi . I agree with your analysis that the current implementation for `has_objects` is incorrect, since it relies on `list_objects` which according to [the documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html?highlight=list_objects#S3.Client.list_objects) indeed returns (by default) only 1000 entries. I have opened a separate issues #15 for this as it is not addressed in this PR.

The method `RepositoryBackend.list_objects` is called internally by the `erase`, `delete_objects`, `has_objects` and by extension `has_object` and `open` methods. These are therefore currently all broken for large repositories.

> we can possibly find a way to implement list_objects to list all the objects in the bucket, but this may be useless and expensive for the server. Is this feature needed anywhere?

Outside of the repository class itself, the `list_objects` method _is_ actually called by other code in AiiDA. Most notably it is called by the archive export/import and the storage maintenance functionality. We are therefore (currently at least) forced to implement it if we want this functionality to be working.

That being said, some of these internal methods do not have to rely on the method. I agree with your approach of changing the `open` method to simply call `download_fileobj` and catch the error in case it doesn't exist. To answer your question:

> call or not call super().open(key)? In principle I think it should be called, but the additional test on file existence is a waste of communication. If this line is removed this implementation works in the same way and is more efficient.

Yes, we should definitely not call the super. The super just performs the check to see if the file exists. As long as our implementation still raises `FileNotFoundError` if it does not, it is fine to skip the super. So please adjust the PR to remove the super call.

Regarding your changes of the `has_objects` method, I am not sure if this is always more performant since now you are performing a request for each key. What happens when someone passes a large list of keys to the method, you get a huge number of requests. I imagine that depending on the number of objects in the storage and the size of the list passed to `has_objects` there is some cross-over point for which actually calling `list_objects` is _actually_ more efficient. Since in this PR you are essentially fixing the `open` call, maybe revert these changes for now? I think we should maybe just fix the `list_objects` implementation and then limit the usage of `has_objects` whenever possible, as you did for `open` now.

rikigigi commented:
Thank you @sphuber for the explanation. I agree on everything. If the method list_objects is used we should definitevely find a better solution. 
Anyway I rebased on master, removed the modification to has_files, and checked that everything worked as expected for my workflow that triggered this issue.

sphuber commented on src/aiida_s3/repository/s3.py:
```diff
@@ -121,11 +121,13 @@ def open(self, key: str) -> t.Iterator[t.IO[bytes]]:  # type: ignore[override]
+                self._client.download_fileobj(self._bucket_name, key, handle)
+                handle.seek(0)
+                yield handle
+        except botocore.exceptions.ClientError:
+            raise FileNotFoundError(f'object with key `{key}` does not exist.')
```
```suggestion
        except botocore.exceptions.ClientError as exception:
            raise FileNotFoundError(f'object with key `{key}` does not exist.') from exception
```

sphuber commented:
Thanks a lot for your contribution @rikigigi ! Much appreciated
